File: utilities/glib/glib/tests/messages-low-memory.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright © 2022 Emmanuel Fleury <emmanuel.fleury@gmail.com>
# Copyright © 2022 Marco Trevisan <mail@3v1n0.net>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA  02110-1301  USA

""" Integration tests for g_message functions on low-memory. """

import os
import unittest

import taptestrunner
import testprogramrunner


class TestMessagesLowMemory(testprogramrunner.TestProgramRunner):
    """Integration test for checking g_message()’s behavior on low memory.

    This can be run when installed or uninstalled. When uninstalled,
    it requires G_TEST_BUILDDIR and G_TEST_SRCDIR to be set.

    The idea with this test harness is to test if g_message and friends
    assert instead of crashing if memory is exhausted, printing the expected
    error message.
    """

    PROGRAM_NAME = "messages-low-memory"
    PROGRAM_TYPE = testprogramrunner.ProgramType.NATIVE

    def test_message_memory_allocation_failure(self):
        """Test running g_message() when memory is exhausted."""
        result = self.runTestProgram([], should_fail=True)

        if result.info.returncode == 77:
            self.skipTest("Not supported")

        if os.name == "nt":
            self.assertEqual(result.info.returncode, 3)
        else:
            self.assertEqual(result.info.returncode, -6)
        self.assertIn("failed to allocate memory", result.err)


if __name__ == "__main__":
    unittest.main(testRunner=taptestrunner.TAPTestRunner())

File: utilities/glib/glib/tests/messages-low-memory.c
/* Unit tests for gmessages on low-memory
 *
 * Copyright (C) 2022 Marco Trevisan
 *
 * SPDX-License-Identifier: LGPL-2.1-or-later
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this library; if not, see <http://www.gnu.org/licenses/>.
 *
 * Author: Marco Trevisan <marco.trevisan@canonical.com>
 */

#include "config.h"

#include <dlfcn.h>
#include <glib.h>

static gboolean malloc_eom = FALSE;
static gboolean our_malloc_called = FALSE;

#ifdef ENOMEM
/* Wrapper around malloc() which returns `ENOMEM` if the test variable
 * `malloc_eom` is set.
 * Otherwise passes through to the normal malloc() in libc.
 */

void *
malloc (size_t size)
{
  static void *(*real_malloc)(size_t);
  if (!real_malloc)
      real_malloc = dlsym (RTLD_NEXT, "malloc");

  if (malloc_eom)
    {
      our_malloc_called = TRUE;
      errno = ENOMEM;
      return NULL;
    }

  return real_malloc (size);
}
#endif

int
main (int   argc,
      char *argv[])
{
  /* We expect this test to abort, so try to avoid that creating a coredump */
  g_test_disable_crash_reporting ();

  g_setenv ("LC_ALL", "C", TRUE);

#ifndef ENOMEM
  g_message ("ENOMEM Not defined, test skipped");
  return 77;
#endif

  g_message ("Simulates a situation in which we were crashing because "
             "of low-memory, leading malloc to fail instead of aborting");
  g_message ("bug: https://gitlab.gnome.org/GNOME/glib/-/issues/2753");

  /* Setting `malloc_eom` to true should cause the override `malloc()`
   * in this file to fail on the allocation on the next line. */
  malloc_eom = TRUE;
  g_message ("Memory is exhausted, but we'll write anyway: %u", 123);

#ifndef __linux__
  if (!our_malloc_called)
    {
      /* For some reasons this doesn't work on darwin systems, so ignore the result
       * for non-linux, while we want to ensure the test is valid at least there
       */
      g_message ("Our malloc implementation has not been called, the test "
                 "has not been performed");
      return 77;
    }
#endif

  return 0;
}
